File: backend/app/schemas/document.py
from datetime import datetime

from pydantic import BaseModel


class DocumentFieldSchema(BaseModel):
    field_key: str
    field_value: str | None = None
    confidence: float | None = None
    page_number: int | None = None

    model_config = {"from_attributes": True}


class DocumentTableSchema(BaseModel):
    table_index: int
    page_number: int | None = None
    headers: list[str] | None = None
    rows: list[list[str]] | None = None

    model_config = {"from_attributes": True}


class DocumentUploadItem(BaseModel):
    id: int
    filename: str
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class DocumentUploadResponse(BaseModel):
    documents: list[DocumentUploadItem]


class DocumentListItem(BaseModel):
    id: int
    filename: str
    original_extension: str
    type: str | None = None
    upload_date: datetime | None = None
    page_count: int | None = None
    status: str

    model_config = {"from_attributes": True}


class DocumentListResponse(BaseModel):
    documents: list[DocumentListItem]
    total: int
    page: int
    per_page: int
    total_pages: int


class DocumentDetail(BaseModel):
    id: int
    filename: str
    original_extension: str
    mime_type: str | None = None
    type: str | None = None
    upload_date: datetime | None = None
    page_count: int | None = None
    status: str
    blob_url: str | None = None
    extracted_text: str | None = None
    fields: list[DocumentFieldSchema] = []
    tables: list[DocumentTableSchema] = []
    error_message: str | None = None

    model_config = {"from_attributes": True}


class ReprocessResponse(BaseModel):
    id: int
    status: str
    message: str

File: frontend/src/types/document.ts
export type DocumentStatus = "uploaded" | "processing" | "processed" | "failed";

export interface Document {
  id: number;
  filename: string;
  original_extension: string;
  upload_date: string;
  page_count: number;
  status: DocumentStatus;
}

export interface ExtractedField {
  field_key: string;
  field_value: string;
  confidence: number;
  page_number: number;
}

export interface ExtractedTable {
  table_index: number;
  page_number: number;
  headers: string[];
  rows: string[][];
}

export interface DocumentDetail extends Document {
  mime_type: string;
  blob_url: string;
  extracted_text: string;
  fields: ExtractedField[];
  tables: ExtractedTable[];
  error_message: string | null;
}

export interface PaginatedDocuments {
  documents: Document[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface UploadResponse {
  documents: {
    id: number;
    filename: string;
    status: string;
    created_at: string;
  }[];
}

export interface DocumentProcessingJobResponse {
  id: string;
  document_id: number;
  status: "PENDING" | "PROCESSING" | "COMPLETED" | "FAILED";
  attempts: number;
  error_log: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface ReprocessResponse {
  id: number;
  status: string;
  message: string;
}
